### PR TITLE
add metadata param for workflow variables

### LINKED ISSUE #298: Update dor-workflow-client to reflect API changes in workflow-server-rails
See sul-dlss/workflow-server-rails#745 for changes this is now including

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Fixes #298 

~~HOLD until workflow-server-rails changes are merged and deployed.~~

Allow the user to send in optional workflow metadata variables.

Note that the response from workflow-server-rails for workflows with metadata just adds to the existing XML.  We add a new method to parse out the metadata for any step into a hash for the consumer (it comes in via the XML response as JSON, which is what is stored in the workflow-server-rails database).

Requires sul-dlss/workflow-server-rails#745 to be released 

Also:
- add the ability to test gem on the console (we do this in other gems too)
- add the missing 'all_workflows' route to base client and a spec for it

## How was this change tested? 🤨

1. New specs
2. Tested on localhost against local workflow-server-rails running new workflow-server variables work

### CODE REVIEW

File: spec/dor/workflow/client/workflow_routes_spec.rb [hunk 1]
```diff
@@ -169,8 +169,48 @@
   end
 
   describe '#create_workflow_by_name' do
-    it 'need to write these specs', skip: 'need to write specs' do
-      # something
+    let(:mock_requestor) { instance_double(Dor::Workflow::Client::Requestor, request: nil) }
```
Comment (peetucket) [inline spec/dor/workflow/client/workflow_routes_spec.rb:172]: hey, i wrote some tests for the default and new scenarios

File: spec/dor/workflow/client_spec.rb [hunk 2]
```diff
@@ -225,6 +225,29 @@
...
+          [200, {}, xml]
+        end
+      end
+    end
+
+    it 'returns the workflow details associated with druid' do
+      expect(client.all_workflows(pid: @druid).xml).to eq(xml)
+    end
+  end
+
```
Comment (peetucket) [inline spec/dor/workflow/client_spec.rb:250]: test for new delegated method added to the client

File: .gitignore [hunk 3]
```diff
@@ -15,5 +15,6 @@ spec/reports
 test/tmp
 test/version_tmp
 tmp
+workflow_service.log
 .ruby-version
```
Comment (peetucket) [inline .gitignore:19]: you will get this file if you test locally on the bin/console script, and you don't want to commit it

File: lib/dor/workflow/response/process.rb [hunk 4]
```diff
@@ -51,6 +51,12 @@ def lane_id
           @attributes[:laneId].presence
         end
 
+        def metadata
+          return nil unless @attributes[:metadata].present?
+
+          JSON.parse(@attributes[:metadata])
+        end
+
         delegate :pid, :workflow_name, to: :parent
```
Comment (peetucket) [inline lib/dor/workflow/response/process.rb:60]: fetch and parse the metadata out of the response, and convert a hash

File: lib/dor/workflow/client/lifecycle_routes.rb [hunk 5]
```diff
@@ -11,7 +11,6 @@ def initialize(requestor:)
 
         # Returns the Date for a requested milestone from workflow lifecycle
         #
-        # @param [String] repo The repository the object resides in. This parameter is deprecated
```
Comment (peetucket) [inline lib/dor/workflow/client/lifecycle_routes.rb:14]: this "repo" parameter is not available on any of these methods anymore

File: spec/dor/workflow/client/workflow_routes_spec.rb (hunk 1 above)
Comment (edsu) [reply spec/dor/workflow/client/workflow_routes_spec.rb:172]: TDD is not dead!

File: spec/dor/workflow/client/workflow_routes_spec.rb [hunk 6]
```diff
@@ -169,8 +169,48 @@
...
+      subject(:create_workflow_by_name) do
+        routes.create_workflow_by_name('druid:mw971zk1113', 'accessionWF', version: '1', lane_id: 'hamburgers')
+      end
+
+      it 'sends a create request without any context param' do
+        create_workflow_by_name
+        expect(mock_requestor).to have_received(:request)
+          .with('objects/druid:mw971zk1113/workflows/accessionWF', 'post', '',
+                { content_type: 'application/json',
+                  params: { 'lane-id' => 'hamburgers', 'version' => '1' } })
```
Comment (edsu) [inline spec/dor/workflow/client/workflow_routes_spec.rb:198]: 🍔

File: spec/dor/workflow/response/process_spec.rb [hunk 7]
```diff
@@ -62,4 +62,32 @@
...
+    context 'when no context exists' do
+      let(:xml) do
+        <<~XML
+          <workflow repository="dor" objectId="druid:mw971zk1113" id="assemblyWF">
+            <process name="start-assembly" laneId="default" context="">
+          </workflow>
+        XML
+      end
+
+      it { is_expected.to be_nil }
```
Comment (edsu) [inline spec/dor/workflow/response/process_spec.rb:90]: Is there any value in having context always be a hash, and either it is populated with things or not?

**Review by edsu — APPROVED** [review]
I left a question, but this looks good, and I see support is already available in the server: https://github.com/sul-dlss/workflow-server-rails/pull/745

I guess we'll need to cut a new release?

File: spec/dor/workflow/response/process_spec.rb (hunk 7 above)
Comment (peetucket) [reply spec/dor/workflow/response/process_spec.rb:90]: Sure, we could do that - no preference either way.

**Comment by peetucket** [comment]
> I left a question, but this looks good, and I see support is already available in the server: [sul-dlss/workflow-server-rails#745](https://github.com/sul-dlss/workflow-server-rails/pull/745)
> 
> I guess we'll need to cut a new release?

Correct, it's ready to be utilized, so we need a new release.

File: spec/dor/workflow/response/process_spec.rb (hunk 7 above)
Comment (peetucket) [reply spec/dor/workflow/response/process_spec.rb]: adjusted to return an empty hash instead of nil when no context available

**Pull request merged**